### ISSUE TITLE
Implement a disk quota mechanism

### ISSUE THREAD

**Issue by prioux** [aces/cbrain#1217]
We need to implement limits for disk space. This quota mechanism would be separate from any mechanism that exists at the file level.

Limits should be associated with:

* DataProviders
* Users
* Groups (aka projects)

and any pair of these things.

Quotas can be set by the admins only. Default quotas must be configurable, with more specific quotas taking priority.

Enforcement would be done at the DataProvider level, when creating new files.  The DB is fast enough to compute something like

```ruby
Userfile.where(:data_provider_id => 1, :user_id => 2, group_id => 3).sum(:size)
```

dynamically every time a new file is about to be created, and compare the value to a set quota.

A new exception class (or several new exception classes) would need to be implemented.

UserKeyDataProviders can never have quotas.

Question: do we only limit the disk space, or do we also create limits for number of files?

**Comment by prioux** [comment]
Default quotas could be configured also in the AccessProfile model, so that when creating an account and associating an AccessProfile to it, account limtis are set. Question: what do do if several AccessProfiles have limits? Or when an AccessProfile is removed from a user?

**Comment by prioux** [comment]
Most of this implemented in #1219 ; missing features (such as AccessProfile integrations) can be added later.
